File: Core/src/main/java/com/inverce/mod/core/threadpool/DynamicScheduledExecutor.java
package com.inverce.mod.core.threadpool;

import android.support.annotation.NonNull;

import java.util.concurrent.Callable;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class DynamicScheduledExecutor extends ThreadPoolExecutor implements ScheduledExecutorService {
    private static ScheduledThreadPoolExecutor scheduler;

    public DynamicScheduledExecutor() {
        super(0, Integer.MAX_VALUE, 1, TimeUnit.SECONDS, new SynchronousQueue<>());
        setThreadFactory(new NamedThreadPool("Dynamic"));
        if (scheduler == null) {
            scheduler = new ScheduledThreadPoolExecutor(1, new NamedThreadPool("Dynamic-Scheduler"));
            scheduler.setKeepAliveTime(1, TimeUnit.SECONDS);
            scheduler.allowCoreThreadTimeOut(true);
        }
    }

    @Override
    public void setKeepAliveTime(long time, TimeUnit unit) {
        super.setKeepAliveTime(time, unit);
        scheduler.setKeepAliveTime(time, unit);
    }

    @NonNull
    @Override
    public ScheduledFuture<?> schedule(@NonNull Runnable command, long delay, @NonNull TimeUnit unit) {
        return scheduler.schedule(wrap(command), delay,  unit);
    }

    @NonNull
    @Override
    public <V> ScheduledFuture<V> schedule(@NonNull Callable<V> callable, long delay, @NonNull TimeUnit unit) {
        return scheduler.schedule(callable, delay, unit);
    }

    @NonNull
    @Override
    public ScheduledFuture<?> scheduleAtFixedRate(@NonNull Runnable command, long initialDelay, long period, @NonNull TimeUnit unit) {
        return scheduler.scheduleAtFixedRate(wrap(command), initialDelay, period, unit);
    }

    @NonNull
    @Override
    public ScheduledFuture<?> scheduleWithFixedDelay(@NonNull Runnable command, long initialDelay, long delay, @NonNull TimeUnit unit) {
        return scheduler.scheduleWithFixedDelay(wrap(command), initialDelay, delay, unit);
    }

    private Runnable wrap(Runnable runnable) {
        //noinspection Convert2Lambda
        return new Runnable() {
            @Override
            public void run() {
                execute(runnable);
            }
        };
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/threadpool/DynamicScheduledExecutor.kt
package com.inverce.mod.v2.core.threadpool

import java.util.concurrent.*

class DynamicScheduledExecutor : ThreadPoolExecutor(0, Integer.MAX_VALUE, 1, TimeUnit.SECONDS, SynchronousQueue()), ScheduledExecutorService {
    protected companion object {
        @set:Synchronized
        @get:Synchronized
        private var count = 0
    }
    protected val scheduler: ScheduledThreadPoolExecutor by lazy {
        ScheduledThreadPoolExecutor(1, NamedThreadPool("Dynamic-Scheduler-${count++}")).apply {
            setKeepAliveTime(1100, TimeUnit.MILLISECONDS)
            allowCoreThreadTimeOut(true)
        }
    }

    init {
        threadFactory = NamedThreadPool("Dynamic")
    }

    override fun setKeepAliveTime(time: Long, unit: TimeUnit) {
        super.setKeepAliveTime(time, unit)
        scheduler.setKeepAliveTime(time, unit)
    }

    override fun schedule(command: Runnable, delay: Long, unit: TimeUnit): ScheduledFuture<*> {
        return scheduler.schedule(wrap(command), delay, unit)
    }

    override fun <V> schedule(callable: Callable<V>, delay: Long, unit: TimeUnit): ScheduledFuture<V> {
        return scheduler.schedule(callable, delay, unit)
    }

    override fun scheduleAtFixedRate(command: Runnable, initialDelay: Long, period: Long, unit: TimeUnit): ScheduledFuture<*> {
        return scheduler.scheduleAtFixedRate(wrap(command), initialDelay, period, unit)
    }

    override fun scheduleWithFixedDelay(command: Runnable, initialDelay: Long, delay: Long, unit: TimeUnit): ScheduledFuture<*> {
        return scheduler.scheduleWithFixedDelay(wrap(command), initialDelay, delay, unit)
    }

    private fun wrap(runnable: Runnable): Runnable {
        return Runnable { execute(runnable) }
    }
}